src/components/game/Strategy.js

```diff
@@ -0,0 +1,236 @@
+// Типо интерфейс
+import { drawRules, getRandomFrom } from './const';
+
+class Strategy {
+  constructor() {
+    this.stage = 'completion'
+  }
+
+  execute(param) { }
+}
+
+// Конкретная стратегия (1)
+export class StrategyPreparation extends Strategy {
+  constructor(param) {
+    super();
+    this.player = param.player
+    this.computer = param.computer
+    this.stage = param.stage
+    this.ctx = param.ctx
+    this.mouse = param.mouse
+    this.func = param.func
+  }
+
+  execute(param) {
+    this.player = param.player
+    this.computer = param.computer
+    this.game = param.game
+
+    // Правила
+    drawRules(this.ctx)
+
+    // Счёт
+    this.computer.getScore(520, 250)
+    this.player.getScore(480, 250)
+
+    // Кнопка
+    const a = document.getElementById('Button');
+    function func(event) {
+      param.game.stage = 'play'
+    }
+
+    a.addEventListener('mousedown', () => {
+      param.player.randoming()
+      func()
+      // param?.game?.stage = 'play'
+    });
+
+    // если мышь не над полем - выход
+    if (!this.player.isPointUnder(this.mouse)) {
+      return
+    }
+
+    // массив размеров кораблей
+    const sheepSizes = [4, 3, 3, 2, 2, 2, 1, 1, 1, 1]
+    // текущий размер корабля
+    const sheepSize = sheepSizes[this.player.sheeps.length]
+    // получаем координаты клетки
+    const coordinats = this.player.getCoordinats(this.mouse)
+
+    const sheep = {
+      x: coordinats.x,
+      y: coordinats.y,
+      direct: this.mouse.s ? 0 : 1,
+      size: sheepSize
+    }
+
+    // если корабль вылезает за поле выход
+    if (!this.player.canStay(sheep)) {
+      return
+    }
+
+    this.player.drawer.drawSheep(this.ctx, sheep)
+
+    // добавление корабля, если в текущей итерации прожата левая кнопка мыши
+    if (this.mouse.left && !this.mouse.pleft) {
+      this.player.addSheeps(sheep)
+
+      // проверяем, что выставили все корабли
+      if (this.player.sheeps.length === 10) {
+        this.stage = 'play'
+      }
+    }
+  }
+}
+
+// Конкретная стратегия (2)
+export class StrategyPlay extends Strategy {
+  constructor(param) {
```

Comment:
❗️
Если параметров много, то так
```suggestion
  constructor(params) {
```

Reply:
согласен